### PR TITLE
[1.9] Updated armour calculations for EntityPlayer, changes to ISpecialArmor

### PR DESCRIPTION
**Pull request by WayofTime**
This is in order to fix https://github.com/MinecraftForge/MinecraftForge/issues/2720 and to implement https://github.com/MinecraftForge/MinecraftForge/issues/2571.

First of all, ISpecialArmor was modernized so that it now takes arguments of EntityEquipmentSlot instead of int slot. 

Secondly, ISpecialArmor was updated so that it provides an extra parameter for ArmorProperties: double Armor. This parameter allows armour pieces that implement ISpecialArmor to provide additional vanilla armour points, while still allowing ISpecialArmor to apply a flat reduction to incoming damage.

ISpecialArmor now implements a hybrid damage reduction formula: when calculating the damage reduction from the armour, it first calculates how much damage is reduced using the vanilla calculations. The remaining damage is then reduced by the percentage reduction based on the provided AbsorbRatio of each piece. It then calculates how much each piece of armour contributed to the entire damage reduction and damages each item accordingly.

Also changed ItemArmor so that it's armour variable's name reflected the fact that it gives armour points instead of raw protection - this last part can be removed if necessary.

### CODE REVIEW

**Comment by WayofTime** [comment]
I need to do a check in ISpecialArmor so that it makes sure the slot type is an ARMOR, so please don't merge until I fix that.

File: patches/minecraft/net/minecraft/item/ItemArmor.java.patch [hunk 1]
```diff
@@ -1,14 +1,64 @@
 --- ../src-base/minecraft/net/minecraft/item/ItemArmor.java
 +++ ../src-work/minecraft/net/minecraft/item/ItemArmor.java
-@@ -235,6 +235,8 @@
-         private final int[] field_78049_g;
+@@ -43,7 +43,7 @@
+         }
+     };
+     public final EntityEquipmentSlot field_77881_a;
+-    public final int field_77879_b;
++    public final int armorAmount;
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/item/ItemArmor.java.patch]: no

File: patches/minecraft/net/minecraft/item/ItemArmor.java.patch [hunk 2]
```diff
@@ -1,14 +1,64 @@
...
++            this.armorAmountArray = p_i46796_5_;
+             this.field_78055_h = p_i46796_6_;
+             this.field_185020_j = p_i46796_7_;
+         }
+@@ -250,9 +252,9 @@
+             return ItemArmor.field_77882_bY[p_78046_1_.func_188454_b()] * this.field_78048_f;
+         }
+ 
+-        public int func_78044_b(EntityEquipmentSlot p_78044_1_)
++        public int getArmorAmount(EntityEquipmentSlot armorType)
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/item/ItemArmor.java.patch]: no

File: patches/minecraft/net/minecraft/item/ItemArmor.java.patch (hunk 1 above)
Comment (iTitus) [inline patches/minecraft/net/minecraft/item/ItemArmor.java.patch]: If you want to fix a method/field name, please use the MCP-Bot or its issue tracker.

Comment (WayofTime) [inline patches/minecraft/net/minecraft/item/ItemArmor.java.patch]: Ok, I'll just remove the name change completely.

File: patches/minecraft/net/minecraft/item/ItemArmor.java.patch [hunk 3]
```diff
@@ -1,28 +0,0 @@
...
-+        public Item customCraftingMaterial = null;
- 
-         private ArmorMaterial(String p_i46796_3_, int p_i46796_4_, int[] p_i46796_5_, int p_i46796_6_, SoundEvent p_i46796_7_)
-         {
-@@ -267,7 +269,15 @@
- 
-         public Item func_151685_b()
-         {
--            return this == LEATHER ? Items.field_151116_aA : (this == CHAIN ? Items.field_151042_j : (this == GOLD ? Items.field_151043_k : (this == IRON ? Items.field_151042_j : (this == DIAMOND ? Items.field_151045_i : null))));
-+            switch (this)
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/item/ItemArmor.java.patch]: ....

Comment (WayofTime) [inline patches/minecraft/net/minecraft/item/ItemArmor.java.patch]: No clue what the heck's up - I'm going to redo this somehow

File: src/main/java/net/minecraftforge/common/ISpecialArmor.java [hunk 4]
```diff
@@ -139,11 +156,11 @@ else if (stack.getItem() instanceof ItemArmor && !source.isUnblockable())
...
 
-                    double absorb = damage * prop.AbsorbRatio;
+                    double absorb = damage * prop.AbsorbRatio + (totalArmor > 0 ? absorbedDamage * prop.Armor / totalArmor : 0);
                     if (absorb > 0)
                     {
-                        ItemStack stack = inventory[prop.Slot];
-                        int itemDamage = (int)(absorb / 25D < 1 ? 1 : absorb / 25D);
+                        ItemStack stack = inventory[prop.Slot.getIndex()];
+                        int itemDamage = (int)(absorb < 1 ? 1 : absorb);
                         if (stack.getItem() instanceof ISpecialArmor)
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/common/ISpecialArmor.java]: Issue is how does one claim to take all of the damage? Vanilla does a flat 1/4th of the damage taken {so you can have an empty slot eat 1/4th of the durability lost} 
But we also want the special armor to be able to say 'Im taking 50% of this damage, no matter what any of the other armor says.'

Comment (WayofTime) [inline src/main/java/net/minecraftforge/common/ISpecialArmor.java]: Well, we have two options: one is to have the ISpecialArmor state if its damage is applied before or after the vanilla calculation, and the other would be to have ISpecialArmor have the ability to apply its reduction before AND after the vanilla one. Which might be a headache to the modder, imo.

File: patches/minecraft/net/minecraft/item/ItemArmor.java.patch (hunk 1 above)
Comment (kashike) [inline patches/minecraft/net/minecraft/item/ItemArmor.java.patch]: This field (`field_77879_b`) has already been named, by the way.

**Comment by WayofTime** [comment]
Closing because I need to scorch-earth my fork to fix the patch issue. Will have a fixed PR in ~30m

**Pull request closed without merging**